eZ/Publish/API/Repository/Values/Content/ContentInfo.php

```diff
@@ -144,6 +144,21 @@ class ContentInfo extends ValueObject
     /** @var bool */
     protected $isHidden;
 
+    /** @var \eZ\Publish\API\Repository\Values\ContentType\ContentType */
+    protected $contentType;
```

Comment:
Do we need this? we already have it on content, which is available also on locations.

Unless there are plans to solve the fact that content info only gives you name in main language, the object is kind of useless imho, you'll always need content in practice anyway.

this is why I left any "Info" object to be plain when adding lazy stuff last year, but of course if name is fixed then that changes it.

Reply:
I'd choose between the two, and I'd actually remove it from `Content`. `Content->info->contentType` is very good already.

Reply:
The thinking was that `ContentInfo` has `contentTypeIdentifier`. Therefore it makes sense that it has `contentType` as well. But `ContentType` is _really_ important, and therefore deserves to be on the `Content`, since it is what we use the most.

So in that API layer, it made sense to have both. In the same line, I also have a `_url`, that returns the main url alias.

Reply:
The _one_ thing to keep in mind is the "generated API" (`Article`, `Image`, ...) we have discussed. There is a risk of names conflicts between these properties on `Content` and the fields definitions coming from the `ContentType`.

Reply:
My 2c:
- happy with adding "convenience" getters to Content which are just shortcuts for accessing contentInfo->$something
- I would not add 'magic getters' to access processed fields data to Content, but at least to a subproperty (ie. content->data->Image) - or to a class which wraps Content (ie. what is done in Kaliop Object Wrapper Bundle). This really helps with understanding and avoids the risk of conflicts

Reply:
Hi Gaetano.

> I would not add 'magic getters' to access processed fields data to Content, but rather to a subproperty (ie. content->data->Image) - or add them instead to a class which wraps Content (ie. what is done in Kaliop Object Wrapper Bundle). This really helps with understanding and avoids the risk of conflicts

While I see your point, especially about conflicts, I think I disagree.

In the majority of day-to-day cases (e.g. view templates), fields are what you care about. I think there are more benefits to making metadata (content info, etc) less visible, and fields more visible.

For conflicts, we always have the option to add reserved names on field definitions identifiers.

Reply:
A subtle/controversial issue... :-) Some feedback from real-life experience:
- some metadata fields we do use extensively in both php and twig, as much as field data if not more (creation/modification date, author name, visibility flags, remote_id)
- although typing `content.contentInfo.created` seems straightforward enough, and some IDEs oblige to suggest that, I have seen scores of developers miss it and use instead `content.contentInfo.versionInfo.created` which is worse, because the IDE also suggested that and they were not arsed to understand the difference. So indeed, having a subproperty or not makes a difference
- often classes have 50 fields or more. I am happy to keep metadata nested to a subproperty, but I would like as well not to mix 'raw field' access with 'custom data' access, as in this scenario the chances of conflict are high, and devs might miss the difference in computing and IO cost between accessing a 'raw' field and something calculated on the fly (in the ez4 docs, tpl objects reference, we used to have a column named 'dynamic' that would tell us which fields where best avoided in terms of perfs, and still lots of devs missed it. I'd like this concept to be more in-your-face rather than less)

Reply:
Thank you for the real-life experience feedback.

While I do see why you would often use creation/modification date and author name, could you give real-life examples of regular usage of visibility flags and remote_id ?

About `content.contentInfo.created` vs `content.contentInfo.versionInfo.created`, I'd almost be tempted to make creation and modification date even easier to reach. Maybe they need to be wrapped and immediately, as you almost always need those ?

Now about the last part, could you elaborate and give examples on the 'raw' vs 'on-the-fly' situations you are encountering these days ? I do get the concept, no worries :)